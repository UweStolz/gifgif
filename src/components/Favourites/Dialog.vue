<template>
  <v-dialog
    v-model="syncedShowDialog"
    :full-image-data="syncedFullImageData"
    :image-id="syncedImageId"
    width="600"
    @click:outside="resetProps"
  >
    <v-card>
      <v-card-title>
        <v-spacer />
        <v-btn
          icon
          @click="resetProps"
        >
          <v-icon>{{ icons.mdiClose }}</v-icon>
        </v-btn>
      </v-card-title>
      <v-img
        :src="imageData"
        :lazy-src="imageData"
        max-height="700"
        height="100%"
        width="100%"
      />
      <v-card-actions>
        <v-btn
          :href="imageData"
          large
          icon
        >
          <v-icon
            id="downloadIcon"
          >
            {{ icons.mdiDownload }}
          </v-icon>
        </v-btn>
        <v-btn
          icon
          large
          @click="deleteGifData"
        >
          <v-icon
            id="trashIcon"
          >
            {{ icons.mdiTrashCan }}
          </v-icon>
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import {
  Vue, Component, PropSync, Watch, Emit,
} from 'vue-property-decorator';
import arrayBufferToImage from '@/util/imageHelper';
import {
  mdiClose, mdiDownload, mdiTrashCan,
} from '@mdi/js';

@Component
export default class Dialog extends Vue {
@Watch('syncedFullImageData')
  convertBufferToImage() {
    if (this.syncedFullImageData) {
      this.imageData = this.syncedFullImageData instanceof ArrayBuffer
        ? arrayBufferToImage(this.syncedFullImageData as ArrayBuffer)
        : this.syncedFullImageData as string;
    }
  }

@PropSync('showDialog', { type: Boolean, required: true }) syncedShowDialog!: boolean;

@PropSync('fullImageData', { required: true }) syncedFullImageData!: null|ArrayBuffer|string;

@PropSync('imageId', { required: true }) syncedImageId!: string;

@Emit('delete')
deleteGifData() {
  this.syncedShowDialog = false;
  return this.syncedImageId;
}

resetProps(): void {
  this.syncedShowDialog = false;
  this.syncedFullImageData = null;
}

icons = {
  mdiClose,
  mdiTrashCan,
  mdiDownload,
}

imageData: string = '';
}
</script>

<style scoped>
#trashIcon:hover{
  color: red;
}
#downloadIcon:hover{
  color:green;
}
</style>
