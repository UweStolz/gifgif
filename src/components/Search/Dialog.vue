<template>
  <v-dialog
    v-model="syncedShowDialog"
    :full-image-data="syncedFullImageData"
    :preview-image-data="syncedpreviewImageData"
    :image-id="syncedImageId"
    :width="$vuetify.breakpoint.xsOnly ? '100%' : '60%'"
    @click:outside="resetProps"
  >
    <v-card>
      <v-card-title>
        <v-spacer />
        <v-btn
          icon
          @click="resetProps"
        >
          <v-icon>{{ icons.mdiClose }}</v-icon>
        </v-btn>
      </v-card-title>
      <v-img
        :src="syncedFullImageData"
        :lazy-src="syncedFullImageData"
        :max-height="maxWindowSize"
        contain
        height="100%"
        width="100%"
      />
      <v-card-actions>
        <v-btn
          large
          icon
          @click="saveImage"
        >
          <v-icon id="downloadIcon">
            {{ icons.mdiDownload }}
          </v-icon>
        </v-btn>
        <v-spacer />

        <v-btn
          v-if="rating !== 0"
          large
          icon
          @click="removeGif"
        >
          <v-icon>
            {{ icons.mdiTrashCan }}
          </v-icon>
        </v-btn>
        <v-rating
          v-model="rating"
          length="5"
          :empty-icon="icons.mdiHeartOutline"
          :full-icon="icons.mdiHeart"
          :half-icon="icons.mdiHeartHalfFull"
          hover
          x-large
          color="red"
          background-color="red"
          @input="updateGifRating"
        />
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import {
  Vue, Component, PropSync, Watch,
} from 'vue-property-decorator';
import {
  mdiClose, mdiDownload, mdiHeartOutline, mdiHeart, mdiHeartHalfFull, mdiTrashCan,
} from '@mdi/js';
import { getBlob } from '@/request';
import { saveAs } from 'file-saver';

@Component
export default class Dialog extends Vue {
  @PropSync('showDialog', { type: Boolean, required: true }) syncedShowDialog!: boolean;

  @PropSync('fullImageData', { type: String, required: true }) syncedFullImageData!: string;

  @PropSync('previewImageData', { type: String, required: true }) syncedpreviewImageData!: string;

  @PropSync('imageId', { type: String, required: true }) syncedImageId!: string;

  @Watch('syncedShowDialog')
  async getRating() {
    const gifData = await this.$store.dispatch('getGifData', `ggid-${this.syncedImageId}`);
    if (gifData) { this.rating = gifData.rating || 0; }
  }

  saveImage(): void {
    saveAs(this.syncedFullImageData, `ggid-${this.syncedImageId}`);
  }

  resetProps(): void {
    this.syncedShowDialog = false;
    this.syncedFullImageData = '';
    this.rating = 0;
  }

  // eslint-disable-next-line class-methods-use-this
  async getBlobForGif(imageData: string) {
    const previewGifBlob = await getBlob(imageData);
    return previewGifBlob;
  }

  async updateGifRating(rating: number) {
    if (this.$store.state.fullImageMode) {
      await this.$store.dispatch('setGifData', {
        rating,
        key: `ggid-${this.syncedImageId}`,
        image: await this.getBlobForGif(this.syncedFullImageData),
        preview: await this.getBlobForGif(this.syncedpreviewImageData),
      });
    } else {
      await this.$store.dispatch('setGifData', {
        rating,
        key: `ggid-${this.syncedImageId}`,
        image: this.syncedFullImageData,
        preview: this.syncedpreviewImageData,
      });
    }
    const currentCount = await this.$store.dispatch('getGifCount');
    if (this.gifCount !== currentCount) { this.$store.commit('setGifCount', currentCount); }
  }


  async removeGif() {
    await this.$store.dispatch('removeGifData', `ggid-${this.syncedImageId}`);
    this.rating = 0;
    this.gifCount -= 1;
    this.$store.commit('setGifCount', this.$store.state.gifCount - 1);
  }

  icons = {
    mdiClose,
    mdiDownload,
    mdiHeartOutline,
    mdiHeart,
    mdiHeartHalfFull,
    mdiTrashCan,
  }

  gifCount: number = 0;

  rating: number = 0;

  maxWindowSize: number = window.innerHeight / 1.5;
}
</script>

<style scoped>
#trashIcon:hover {
  color: red;
}
#downloadIcon:hover {
  color: green;
}
</style>
