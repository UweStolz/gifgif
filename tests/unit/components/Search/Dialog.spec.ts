import Dialog from '@/components/Search/Dialog.vue';
import shallow from '../../../helper';

describe('Dialog.vue', () => {
  it('renders properly', () => {
    const wrapper = shallow(Dialog);
    expect(wrapper).toMatchSnapshot();
  });
});
