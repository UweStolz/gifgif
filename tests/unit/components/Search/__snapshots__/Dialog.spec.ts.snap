// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog.vue renders properly 1`] = `
<v-dialog-stub
  closedelay="0"
  contentclass=""
  full-image-data="http://someFullImage.webp"
  image-id="ggid-123321"
  maxwidth="none"
  opendelay="0"
  origin="center center"
  preview-image-data="http://somePreviewImage.webp"
  retainfocus="true"
  transition="dialog-transition"
  width="60%"
>
  <v-card-stub
    loaderheight="4"
    style="overflow: hidden;"
    tag="div"
  >
    <v-card-title-stub>
      <v-spacer-stub />
       
      <v-btn-stub
        activeclass=""
        icon="true"
        tag="button"
        type="button"
      >
        <v-icon-stub>
          M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z
        </v-icon-stub>
      </v-btn-stub>
    </v-card-title-stub>
     
    <v-img-stub
      contain="true"
      height="100%"
      lazysrc="http://someFullImage.webp"
      maxheight="512"
      options="[object Object]"
      position="center center"
      src="http://someFullImage.webp"
      transition="fade-transition"
      width="100%"
    />
     
    <card-actions-stub
      imagedata="http://someFullImage.webp"
      imageid="ggid-123321"
      previewimagedata="http://somePreviewImage.webp"
      rating="0"
      showrating="true"
    />
  </v-card-stub>
</v-dialog-stub>
`;
